### PR TITLE
http: add http server

### PR DESCRIPTION
**Pull request by mmcgee-jump**
Related to #1813

### CODE REVIEW

File: src/ballet/http/fd_http_server.c [hunk 1]
```diff
@@ -213,21 +204,34 @@ fd_http_server_listen( fd_http_server_t * http,
...
+  FD_LOG_NOTICE(( "listening to port %hu", port ));
+
   return http;
 }
 
 static void
 close_conn( fd_http_server_t * http,
             ulong              conn_idx,
             int                reason ) {
+  if( http->pollfds[ conn_idx ].fd == -1 ) return;
```
Comment (mmcgee-jump) [inline src/ballet/http/fd_http_server.c:216]: Not sure why this would happen?

File: src/ballet/http/fd_http_server.c [hunk 2]
```diff
@@ -251,8 +255,14 @@ accept_conns( fd_http_server_t * http ) {
     http->conns[ http->conn_id ].state                  = FD_HTTP_SERVER_CONNECTION_STATE_READING;
     http->conns[ http->conn_id ].request_bytes_read     = 0UL;
     http->conns[ http->conn_id ].response_bytes_written = 0UL;
+    memset( &http->conns[ http->conn_id ].response, 0, sizeof(fd_http_server_response_t) );
     http->conn_id = (http->conn_id+1UL) % http->max_conns;
 
+    int optval = 2<<20;
```
Comment (mmcgee-jump) [inline src/ballet/http/fd_http_server.c:261]: Make a server parameter, optionally provided by caller

File: src/ballet/http/fd_http_server.c [hunk 3]
```diff
@@ -651,7 +715,7 @@ maybe_write_pong( fd_http_server_t * http,
 
   long sz = write( http->pollfds[ conn_idx ].fd, frame+conn->pong_bytes_written, 2UL+conn->pong_data_len-conn->pong_bytes_written );
   if( FD_UNLIKELY( -1==sz && (errno==EAGAIN || errno==EINTR) ) ) return 1; /* No data was written, continue. */
-  else if( FD_UNLIKELY( -1==sz && (errno==EPIPE || errno==ECONNRESET) ) ) {
+  else if( FD_UNLIKELY( -1==sz && (errno==EPIPE || errno==ECONNRESET || errno==EBADF) ) ) {
```
Comment (mmcgee-jump) [inline src/ballet/http/fd_http_server.c:718]: Why? EBADF is programmer error?

File: src/ballet/http/fd_http_server.c [hunk 4]
```diff
@@ -473,23 +550,6 @@ read_conn_ws( fd_http_server_t * http,
...
 
-  /* Data frame, process it */
-
-  int is_fin_set = conn->recv_bytes[ conn->recv_bytes_parsed+0UL ] & 0x80;
-
-  uchar * mask    = conn->recv_bytes+conn->recv_bytes_parsed+1UL+len_bytes;
-  uchar * payload = conn->recv_bytes+conn->recv_bytes_parsed+header_len;
-  for( ulong i=0UL; i<payload_len; i++ ) conn->recv_bytes[ conn->recv_bytes_parsed+i ] = payload[ i ] ^ mask[ i % 4 ];
-
-  /* Check if this is a complete message */
```
Comment (mmcgee-jump) [inline src/ballet/http/fd_http_server.c]: This needs to stay down here

File: src/ballet/http/fd_http_server.c [hunk 5]
```diff
@@ -213,21 +204,34 @@ fd_http_server_listen( fd_http_server_t * http,
...
   if( FD_LIKELY( conn_idx<http->max_conns ) ) {
-    if( FD_LIKELY( http->callbacks.close    ) ) http->callbacks.close( conn_idx, reason, http->callback_ctx );
+    struct fd_http_server_connection * conn = &http->conns[ conn_idx ];
+    if( FD_LIKELY( http->callbacks.close    ) ) http->callbacks.close( conn_idx, reason, &conn->response, http->callback_ctx );
+
   } else {
+    struct fd_http_server_ws_connection * conn = &http->ws_conns[ conn_idx-http->max_conns ];
+    while( conn->send_frame_cnt ) {
+      fd_http_server_ws_frame_t * frame = &conn->send_frames[ conn->send_frame_idx ];
+      if( FD_LIKELY( http->callbacks.ws_sent ) ) http->callbacks.ws_sent( conn_idx-http->max_conns, frame, http->callback_ctx );
```
Comment (mmcgee-jump) [inline src/ballet/http/fd_http_server.c:231]: This isn't `ws_sent`? The frame was dropped due to connection close

Comment (asiegel-jt) [reply src/ballet/http/fd_http_server.c]: Changed to ws_done

File: src/ballet/http/fd_http_server.c (hunk 1 above)
Comment (asiegel-jt) [reply src/ballet/http/fd_http_server.c]: I made it a FD_TEST

File: src/ballet/http/fd_http_server.c [hunk 6]
```diff
@@ -213,21 +204,34 @@ fd_http_server_listen( fd_http_server_t * http,
   http->socket_fd = sockfd;
   http->pollfds[ http->max_conns+http->max_ws_conns ].fd = http->socket_fd;
 
+  FD_LOG_NOTICE(( "listening to port %hu", port ));
```
[the remark this answers is not in the document]
Comment (asiegel-jt) [reply src/ballet/http/fd_http_server.c]: Done

**Review by asiegel-jt — DISMISSED** [review]

File: src/ballet/http/fd_hcache.h [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+   The caller should return the body data as a response body to an HTTP
+   request immediately, as this is how the hcache manages the lifetime
+   of the buffer.  If the buffer is not being used to send a response
+   to any connected HTTP client, it could be overwritten by the next
+   snap and become no longer valid.
+   
+   Returns NULL on failure if the hcache is in an error state because
+   the message being snapped exceeds the entire hcache data size. */
+
```
Comment (asiegel-jt) [inline src/ballet/http/fd_hcache.h:161]: I also need a fd_hcache_reset method which sets snap_len to zero. This is useful if we need to switch to an error response mid-stream

**Comment by asiegel-jt** [comment]
test_sha1 is failing

File: src/ballet/http/fd_hcache.h [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+   so far and reset the hcache to start preparing the next message.
+
+   Copying is not error-free, it is assumed that the data is valid but
+   the entire data region may not be large enough to hold the copied
+   data.  In that case, the hcache is marked as being in an error state
+   internally.  The next call to fd_hcache_snap_ws_broadcast will fail
+   and the error state will be cleared. */
+
+void
+fd_hcache_memcpy( fd_hcache_t * hcache,
```
Comment (asiegel-jt) [inline src/ballet/http/fd_hcache.h:144]: You need to actually implement this

**Review by asiegel-jt — DISMISSED** [review]

**Review by asiegel-jt — APPROVED** [review]

**Pull request merged**